### PR TITLE
[GH-2799] Add ST_OffsetCurve function

### LINKED ISSUE #2799: Add ST_OffsetCurve function
## Summary

Add `ST_OffsetCurve(geom, distance [, quadrantSegments])` function that returns a line offset by a given distance from a linear geometry.

This is the PostGIS equivalent of [`ST_OffsetCurve`](https://postgis.net/docs/ST_OffsetCurve.html).

**Parent issue:** #2230

## Design

JTS 1.19+ provides `org.locationtech.jts.operation.buffer.OffsetCurve` which computes an offset curve for a linear geometry. Sedona uses JTS 1.20.0, so this class is available.

- Positive distance → offset to the left side
- Negative distance → offset to the right side

### Signature

```sql
ST_OffsetCurve(geometry lineGeom, double distance) → geometry (LineString)
ST_OffsetCurve(geometry lineGeom, double distance, int quadrantSegments) → geometry (LineString)
```

### Implementation layers

1. **Java core** (`Functions.java`): Wrapper around `new OffsetCurve(geometry, distance).getCurve()`; overload with `quadrantSegments` using `OffsetCurve(geometry, distance, quadSegs).getCurve()`
2. **Scala Expression** (`Functions.scala`): `InferredExpression` case class with multiple overloads
3. **Catalog registration** (`Catalog.scala`)
4. **Scala API** (`st_functions.scala`): Column wrappers
5. **Python API** (`st_functions.py`): Python wrapper
6. **Tests**: Scala + Python

### Example

```sql
SELECT ST_OffsetCurve(
  ST_GeomFromWKT('LINESTRING(0 0, 10 0)'),
  5.0
);
-- LINESTRING (0 5, 10 5)
```

## Edge cases

- Non-linear geometries: should handle gracefully (return null or throw)
- Zero distance: should return a copy of the input
- Empty geometries: return null

### PR DESCRIPTION
**Pull request by jiayuasu**
## Summary

- Add `ST_OffsetCurve(geom, distance [, quadrantSegments])` function that returns a line offset by a given distance from a linear geometry
- Uses JTS `org.locationtech.jts.operation.buffer.OffsetCurve` (available since JTS 1.19, Sedona uses 1.20)
- Positive distance offsets to the left, negative to the right
- Returns `null` for empty geometries

Closes #2799

### Implementation (6 layers)
- **Java core** (`Functions.java`): Two overloads — `offsetCurve(Geometry, double)` and `offsetCurve(Geometry, double, int)` using JTS `OffsetCurve` with `BufferParameters`
- **Scala Expression** (`Functions.scala`): `ST_OffsetCurve` case class with `inferrableFunction2` and `inferrableFunction3`
- **Catalog** (`Catalog.scala`): Function registration
- **Scala API** (`st_functions.scala`): Column/String wrappers with optional quadrantSegments
- **Python API** (`st_functions.py`): Python wrapper with optional `quadrant_segments` parameter
- **Flink** (`Functions.java`, `Catalog.java`): Flink scalar function + registration

### Tests
- Java unit tests (5 cases: positive offset, negative offset, zero distance, empty geometry, custom quadrant segments)
- Scala SQL tests + DataFrame API tests (with and without quadrantSegments)
- SRID preservation test
- Python SQL test + DataFrame API tests
- Flink tests (2 cases)

### Documentation
- SQL, Flink, and Snowflake docs with examples and SVG visual

## Test plan

- [x] Java unit tests pass
- [x] Python SQL tests pass
- [x] Python DataFrame API tests pass
- [ ] CI full test suite

### CODE REVIEW

File: docs/api/sql/Geometry-Processing/ST_OffsetCurve.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+![ST_OffsetCurve](../../../image/ST_OffsetCurve/ST_OffsetCurve.svg "ST_OffsetCurve")
+
+Introduction: Returns a line at a given offset distance from a linear geometry. If the distance is positive, the offset is on the left side of the input line; if it is negative, it is on the right side. Returns null for empty geometries.
+
+The optional third parameter `quadrantSegments` controls the number of line segments used to approximate a quarter circle at round joins. The default value is 8.
+
+Format: `ST_OffsetCurve(geometry: Geometry, distance: Double, quadrantSegments: Integer)`
+
+Format: `ST_OffsetCurve(geometry: Geometry, distance: Double)`
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply docs/api/sql/Geometry-Processing/ST_OffsetCurve.md:28]: Consolidated into a single optional-parameter format and switched to fenced output blocks.

File: docs/api/flink/Geometry-Processing/ST_OffsetCurve.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+Introduction: Returns a line at a given offset distance from a linear geometry. If the distance is positive, the offset is on the left side of the input line; if it is negative, it is on the right side. Returns null for empty geometries.
+
+The optional third parameter `quadrantSegments` controls the number of line segments used to approximate a quarter circle at round joins. The default value is 8.
+
+Format: `ST_OffsetCurve(geometry: Geometry, distance: Double, quadrantSegments: Integer)`
+
+Format: `ST_OffsetCurve(geometry: Geometry, distance: Double)`
+
+Return type: `Geometry`
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply docs/api/flink/Geometry-Processing/ST_OffsetCurve.md:30]: Same fix applied here.

File: common/src/main/java/org/apache/sedona/common/Functions.java [hunk 3]
```diff
@@ -424,6 +425,20 @@ else if (singleParam[0].equalsIgnoreCase(listBufferParameters[5])) {
...
+  }
+
+  public static Geometry offsetCurve(Geometry geometry, double distance, int quadrantSegments) {
+    if (geometry.isEmpty()) {
+      return null;
+    }
+    BufferParameters bufferParameters = new BufferParameters();
+    bufferParameters.setQuadrantSegments(quadrantSegments);
+    OffsetCurve oc = new OffsetCurve(geometry, distance, bufferParameters);
+    return oc.getCurve();
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply common/src/main/java/org/apache/sedona/common/Functions.java:439]: Spark/Flink expressions null-short-circuit before reaching these Java functions. The existing codebase is inconsistent about this — some functions have the null guard, some don't. Not changing it here to stay consistent with the surrounding code.

File: snowflake-tester/src/test/java/org/apache/sedona/snowflake/snowsql/TestFunctions.java [hunk 4]
```diff
@@ -768,6 +768,19 @@ public void test_ST_NPoints() {
...
+  public void test_ST_OffsetCurve() {
+    registerUDF("ST_OffsetCurve", byte[].class, double.class);
+    verifySqlSingleRes(
+        "select sedona.ST_AsText(sedona.ST_OffsetCurve(sedona.ST_GeomFromText('LINESTRING(0 0, 10 0)'), 5.0))",
+        "LINESTRING (0 5, 10 5)");
+    registerUDF("ST_OffsetCurve", byte[].class, double.class, int.class);
+    registerUDF("ST_NPoints", byte[].class);
+    verifySqlSingleRes(
+        "select sedona.ST_NPoints(sedona.ST_OffsetCurve(sedona.ST_GeomFromText('LINESTRING(0 0, 10 0, 10 10)'), -3.0, 16))",
+        19);
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply snowflake-tester/src/test/java/org/apache/sedona/snowflake/snowsql/TestFunctions.java:781]: Changed to a relative comparison using Snowflake's `iff()` to assert qs=16 produces more points than the default.

File: snowflake-tester/src/test/java/org/apache/sedona/snowflake/snowsql/TestFunctionsV2.java [hunk 5]
```diff
@@ -716,6 +716,19 @@ public void test_ST_NPoints() {
...
+  public void test_ST_OffsetCurve() {
+    registerUDFV2("ST_OffsetCurve", String.class, double.class);
+    verifySqlSingleRes(
+        "select sedona.ST_AsText(sedona.ST_OffsetCurve(ST_GeometryFromWKT('LINESTRING(0 0, 10 0)'), 5.0))",
+        "LINESTRING (0 5, 10 5)");
+    registerUDFV2("ST_OffsetCurve", String.class, double.class, int.class);
+    registerUDFV2("ST_NPoints", String.class);
+    verifySqlSingleRes(
+        "select sedona.ST_NPoints(sedona.ST_OffsetCurve(ST_GeometryFromWKT('LINESTRING(0 0, 10 0, 10 10)'), -3.0, 16))",
+        19);
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply snowflake-tester/src/test/java/org/apache/sedona/snowflake/snowsql/TestFunctionsV2.java:729]: Same fix applied here.
